src/shared/dissect-image.c

```diff
@@ -2004,10 +2004,25 @@ static int mount_partition(
 
         if (p) {
                 if (m->fsmount_fd >= 0) {
+                        bool mount_beneath = FLAGS_SET(flags, DISSECT_IMAGE_MOUNT_BENEATH);
+
                         /* Case #1: Attach existing fsmount fd to the file system */
 
-                        if (move_mount(m->fsmount_fd, "", -EBADF, p, MOVE_MOUNT_F_EMPTY_PATH) < 0)
-                                return -errno;
+                        r = RET_NERRNO(move_mount(m->fsmount_fd,
+                                       "",
+                                       -EBADF,
+                                       p,
+                                       MOVE_MOUNT_F_EMPTY_PATH | (mount_beneath ? MOVE_MOUNT_BENEATH : 0)));
+                        if (mount_beneath && r == -EINVAL) /* Fallback if mount_beneath is not supported */
+                                r = RET_NERRNO(move_mount(m->fsmount_fd,
+                                               "",
+                                               -EBADF,
+                                               p,
+                                               MOVE_MOUNT_F_EMPTY_PATH));
+                        else if (mount_beneath && r >= 0) /* If it is, now remove the old mount */
+                                r = RET_NERRNO(umount2(p, UMOUNT_NOFOLLOW|MNT_DETACH));
+                        if (r < 0)
+                                return r;
 
                 } else {
                         assert(node);
```

Comment:
so if you have a DDI mounted that consists of multiple partitions, i.e. root, /usr, /srv, and then you use this to replace it with another DDI, i presume this is not going to work at all and is going to fail with EBUSY or something, right? because MOUNT_BENEATH doesn't allow mounting beneath mount points that have child mounts, right?

I am pretty sure this case should be handled, because otherwise this would be a regression: as before we should have the DDI mounted in the end, though in the special cases where we can do unmount the underlying mount we should

Reply:
The intention was to fallback to an overmount, but let me check that I got the errnos right

Reply:
The man from the mountain said it actually works, there's no -ebusy, so I've added a comment to clarify this.